tokio/src/sync/semaphore.rs

```diff
@@ -102,6 +102,79 @@ use std::sync::Arc;
 /// }
 /// ```
 ///
+/// Implement a simple token bucket for rate limiting
+///
+/// Many applications and systems have constraints on the rate at which certain
+/// operations should occur. Exceeding this rate can result in suboptimal
+/// performance or even errors.
+///
+/// The provided example uses a `TokenBucket`, implemented using a semaphore, that
+/// limits operations to a specific rate although token buckets allow short bursts that are faster
+/// than the allowed rate. The token bucket will be refilled after each interval. When the rate is exceeded,
+/// the `acquire` method will await until a token is available. Note that this implementation is suboptimal
+/// when the rate is large, because it consumes a lot of cpu constantly looping and sleeping.
+/// ```
+/// use std::sync::Arc;
+/// use tokio::sync::{AcquireError, Semaphore};
+/// use tokio::time::{interval, Duration};
+///
+/// struct TokenBucket {
+///     sem: Arc<Semaphore>,
+///     jh: tokio::task::JoinHandle<()>,
+/// }
+///
+/// impl TokenBucket {
+///     fn new(duration: Duration) -> Self {
+///         let rate = (1.0 / duration.as_secs_f32()) as usize;
```

Comment:
You're only using this to compute the maximum capacity. Instead, I suggest making the capacity into a separate argument.

Reply:
Well the capacity is calculated based on the `duration`. If one specifies the rate, the duration is determined, and visa versa. Do you mean two `new` functions, one with `duration` and one with `rate` as an argument?

Reply:
No, I mean one function with both parameters. Token buckets usually don't compute the capacity based on the duration.